src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -598,18 +598,30 @@ size_t TieredHNSWIndex<DataType, DistType>::indexLabelCount() const {
 template <typename DataType, typename DistType>
 int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType label,
                                                    void *auxiliaryCtx) {
-
-    if (this->getWriteMode() == VecSim_WriteInPlace ||
-        this->frontendIndex->indexSize() >= this->flatBufferLimit) {
-        auto hnsw_index = this->getHNSWIndex();
-        // Insert vector directly to HNSW (since flat buffer guard was not held, no need to release
-        // it internally).
-        this->insertVectorToHNSW<false>(hnsw_index, label, blob);
+    int ret = 1;
+    auto hnsw_index = this->getHNSWIndex();
+    if (this->getWriteMode() == VecSim_WriteInPlace) {
+        this->mainIndexGuard.lock();
+        // Internally, we may increase the capacity when we append thr new vector,
+        ret = hnsw_index->addVector(blob, label);
+        this->mainIndexGuard.unlock();
         this->UpdateIndexMemory(this->memoryCtx, this->getAllocationSize());
-        return 1;
+        return ret;
+    }
+    if (this->frontendIndex->indexSize() >= this->flatBufferLimit) {
+        // Handle overwrite situation.
+        if (!this->backendIndex->isMultiValue()) {
+            ret -= this->deleteVector(label);
+        }
+        if (this->frontendIndex->indexSize() >= this->flatBufferLimit) {
```

Comment:
this second check is redundant

Reply:
It doesn't, since `this->deleteVector(label)` may decrease the flat buffer size...

Reply:
Right...